Support cancellation of batch compilation

retronym opened the pull request:
A common means of cancelling a task is to shutdown the thread pool
executing it. That's what SBT's CTRL-C handler does, for example.
Typically, thread pools call `Thread.interrupt()` to cooperatively stop the
workload. We need to do our part by checking `interrupted()` from time to
time, and translating this into an exception that will stop compilation.

Side note: SBT's CTRL-C handler is not on by default, you need to add
the following to `build.sbt`.

```scala
cancelable in Global := true
```

We actually were doing this in some places, by accident, by virtue
of calling NIO file reading/writing APIs, which throw a specific exception
if the thread is interrupted. The first commits here harden the places
where we were preventing these exceptions from bubbling up and stopping the
compiler. Before this change, an interrupt during the parser or bcode phase
ended up issuing an stack trace and/or an error for each source/class file.

The final commit adds additional checks for the interrupt status. Perhaps the
check within `Typer.typed` is contentious, it aims to make cancellation
responsive even in large source files. We'll see if there is a performance
difference here, and perhaps move it to `typedDefDef` where it will fire
with an appropriate frequency.

Tested by using these commits as STARR for our SBT build and hitting
CTRL-C at different moments during `library/compile`.

## baseline behaviour

### Cancel during parser:
```
[error] IO error while decoding /Users/jz/code/scala/src/library/scala/Tuple5.scala with UTF-8
[error] Please try specifying another one using the -encoding option
[error] IO error while decoding /Users/jz/code/scala/src/library/scala/Tuple6.scala with UTF-8
[error] Please try specifying another one using the -encoding option
[error] IO error while decoding /Users/jz/code/scala/src/library/scala/Tuple7.scala with UTF-8
[error] Please try specifying another one using the -encoding option
[error] 704 errors found
```

### Cancel after parser

```
> library/compile
[info] Compiling 584 Scala sources and 120 Java sources to /Users/jz/code/scala/build/quick/classes/library...
^C
[warn] Canceling execution...

<BIG PAUSE/>

java.nio.channels.ClosedByInterruptException
    at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
    at sun.nio.ch.FileChannelImpl.writeInternal(FileChannelImpl.java:783)
    at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:764)
    at scala.tools.nsc.backend.jvm.ClassfileWriters$ClassfileWriter$DirClassWriter.write(ClassfileWriters.scala:179)
    at scala.tools.nsc.backend.jvm.PostProcessor.sendToDisk(PostProcessor.scala:76)
...
java.nio.channels.ClosedByInterruptException
    at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
    at sun.nio.ch.FileChannelImpl.writeInternal(FileChannelImpl.java:783)
    at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:764)
    at scala.tools.nsc.backend.jvm.ClassfileWriters$ClassfileWriter$DirClassWriter.write(ClassfileWriters.scala:179)
    at scala.tools.nsc.backend.jvm.PostProcessor.sendToDisk(PostProcessor.scala:76)
...
java.nio.channels.ClosedByInterruptException
    at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
    at sun.nio.ch.FileChannelImpl.writeInternal(FileChannelImpl.java:783)
    at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:764)
    at scala.tools.nsc.backend.jvm.ClassfileWriters$ClassfileWriter$DirClassWriter.write(ClassfileWriters.scala:179)
    at scala.tools.nsc.backend.jvm.PostProcessor.sendToDisk(PostProcessor.scala:76)
...

[error] 584 errors found

```

## New behaviour

### Cancel any time

```
[info] Compiling 584 Scala sources and 120 Java sources to /Users/jz/code/scala/build/quick/classes/library...
^C
[warn] Canceling execution...
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ExecutorCompletionService$QueueingFuture@35798cd0 rejected from java.util.concurrent.ThreadPoolExecutor@2fc70622[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 194]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
```

retronym committed:
Support cancellation by checking Thread.interrupted

A common means of cancelling a task is to shutdown the thread
pool executing it. That's what SBT's CTRL-C handler does, for example.

Typically, thread pools call `Thread.interrupt()` to cooperatively
stop the workload. We need to do our part by checking `interrupted()`
from time to time, and translating this into an exception that will
stop compilation.

retronym committed:
Honour Thread.interrupt detected by NIO operations

NIO file operations check for thread interruption and throw an
exception. If we see one of these, all subsequent operations will
fail in the same way, so we should fail fast.

Since the change to use NIO for file writing, and prior to this change,
hitting CTRL-C during `sbt compile` would spew out a stack trace
for each classfile.

retronym committed:
Honour interrupts in source file reading

retronym committed:
Delete empty classfile if thread is interrupted during writing

retronym commented:
@lrytz Targetting to 2.12.x, perhaps even 2.12.6, because the changes to use NIO in the backend for class writing seem to expose us to the error spew more than before.

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/ClassfileWriters.scala:
```diff
@@ -179,7 +179,17 @@ abstract class ClassfileWriters {
+        } catch {
+          case ex: ClosedByInterruptException =>
+            try {
+              Files.deleteIfExists(path) // don't leave a empty of half-written classfile around after an interrupt
```
I noticed errors on subsequent compiles after compilation like `class file .../package.class is broken / ArrayIndexOutOutBoundsException(0)`. My guess was I was seeing a file that had been `open`-ed but not written to.

retronym replied:
Maybe we need to look more closely at what SBT/Zinc does with its management of classfiles on interruption.

jvican replied:
I think the transactional classfile manager will detect that zinc failed, and restore the class files that were available before compiling. So within Zinc, this should not be a problem.

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/CodeGen.scala:
```diff
@@ -43,6 +43,7 @@ abstract class CodeGen[G <: Global](val global: G) extends PerRunInit {
     } catch {
+      case ex: InterruptedException => throw ex
       case ex: Throwable =>
         ex.printStackTrace()
```
I thought this should probably be guarded with `if (settings.debug)`, @lrytz wdyt?

lrytz replied:
Yes, agree.

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/PostProcessor.scala:
```diff
@@ -63,7 +65,10 @@ abstract class PostProcessor extends PerRunInit {
       case ex: Throwable =>
+        // TODO hide this stack trace behind -Ydebug?
+        // TODO fail fast rather than continuing to write the rest of the class files?
         ex.printStackTrace()
```
Same question here

lrytz replied:
Agree to guard with `if (settings.debug)`. I also agree to fail fast instead of continuing with writing other classfiles.

retronym commented on src/compiler/scala/tools/nsc/io/SourceReader.scala:
```diff
@@ -38,7 +39,13 @@ class SourceReader(decoder: CharsetDecoder, reporter: Reporter) {
+      case ex: InterruptedException => throw ex
+      case _: ClosedByInterruptException => throw new InterruptedException
+      case e: Exception =>
+        e.printStackTrace()
```
Accidentally pushed this change, will remove.

jvican commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -423,7 +423,10 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
 
-    final def applyPhase(unit: CompilationUnit) = withCurrentUnit(unit)(apply(unit))
+    final def applyPhase(unit: CompilationUnit) = {
+      if (Thread.interrupted()) throw new InterruptedException
```
Is it possible that instead of checking `Thread.interrupted` we check a variable in `Global`? Sometimes it's not possible to interrupt execution thread pools, and it's actually a costly operation to shut it down completely and then start it up again for another compilation run.

jvican replied:
If we do this, this would be highly beneficial for tools that control zinc like bloop or pants (/cc @stuhood) because we'd be able to set the variable in `Global` whenever we detect a cancellation in our task engine (in bloop's case, that's Monix).

retronym replied:
The surrounding tool need not shut down the threadpool.

```scala
var compilerThread: Thread = null
var interruped = false
threadPool.submit( () => {
  try {
    compilerThread = Thread.currentThread
    new Global().compile()
  }  catch { 
    case _: InterruptedException =>
      interrupted = false
  }
})

// elsewhere
Option(compilerThread).foreach(_.interrupt)
```

The calls in this PR to `Thread.interrupted` will clear the interrupted bit of the thread, so it is fine to continue work in the thread pool.

jvican replied:
Indeed, but the issue here is that the surrounding tool may not have access to the thread running the compiler. In the case of bloop, we use Monix and we cannot do `interrupt`. If anoter tool was using `Future`, it would have the same problem.

retronym replied:
You can capture the current thread just before you call `compile`, as I did above.

I suppose that, as a convenience, we could have `Run` capture the current thread at the start of in `compileUnitsInternal`, and then have a `Run.cancel()` method that interrupted that thread.

jvican replied:
> You can capture the current thread just before you call compile, as I did above.

Doesn't that assume that it will always compile a program with the thread that starts the compilation? I think that assumption doesn't hold with common execution contexts.

jvican replied:
@retronym ping ^

retronym replied:
The latest implementation checks for both `Reporter.cancelled` and `Thread.interrupted`, so the caller can use whichever is more suitable.

That said, In the near term, it is unlikely that the compiler would transfer the main phase application loop to a different thread from the one that calls `compile`.

jvican reviewed (commented):
I'm really happy to see this improvement in 2.12.x. I was lately thinking of opening a ticket to discuss the best approach to fix this problem. :tada:

retronym commented:
I've just noticed some other code that handles cancellation:

https://github.com/scala/scala/blob/f4ba045ffc4e84c85bd04a045af5adaad5eaa54e/src/compiler/scala/tools/nsc/Global.scala#L396-L413

https://github.com/sbt/zinc/blob/41310b4fccf940de3573bc37ef6ffd1026c95eaa/internal/compiler-bridge/src/main/scala/xsbt/CompilerInterface.scala#L140-L156

retronym commented:
So one way could go would be to change `Global.cancel()` to `Thread.interrupt()`, and just use the new code in this PR to detect cancellation.

The risk of that is that the call happens just before a successful completion of compilation, so we never check `Thread#interrrupted`, and we leave the thread in the interrupted state after a successful compile.

jvican commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5615,6 +5615,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       val statsEnabled = StatisticsStatics.areSomeHotStatsEnabled() && statistics.areHotStatsLocallyEnabled
       val startByType = if (statsEnabled) statistics.pushTimer(byTypeStack, byTypeNanos(tree.getClass)) else null
       if (statsEnabled) statistics.incCounter(visitsByType, tree.getClass)
+      if (Thread.interrupted()) throw new InterruptedException
```
This is the magic to make it work in most of the cases :smile:. Have you benchmarked it?

mkeskells replied:
No- dont do this - check can be expensive

retronym replied:
Removed for now. We could add it back into, say, `typedDefDef` to check far less frequently.

lrytz replied:
It's not yet removed (it seems).

retronym replied:
Removed, we'll go back to for the end-of-compilation unit check.

mkeskells commented:
Interrupting a thread assumed and requires single threaded operation
This implementation assumes that NIO will throw exceptions based on file access (which is optional and depends on FileSystem implementation)
This is fragile an world require significant investment in testing infrastructure (different OS, filesystems), parallelisation options)

Wouldn't a simple solution- checking a flag in the Global.run suffice, between phases, or between CU for expensive phases (typer) by sufficient

Clearly need to cope with the exceptions that are emitted better though

retronym committed:
Combine thread interrupt handling with Reporter.cancelled

Also remove the fine grained cancellation checking within
typechecking a compilation unit.

retronym commented:
The fact is that tools currently use `Thread.interrupt()`, so we either need to change those tools (and wait for people to upgrade) or expand our cancellation API to make it work as they expect. IMO it isn't a big burden our our codebase to  check `Thread.interrupted` (being careful, not to be too fine grained) and handle `ClosedByInterruptException` when calling NIO.

I've pushed a commit that removes the fine-grained checking on every AST node, instead just waiting until the end of the next compilation unit that gets typechecked.

It also integrates better with the existing `Reporter.cancelled` flag, so `Reporter.hasError` will be true after a thread cancellation.

Regarding the multi-threaded support in the backend: if the backend fails fast, either due to cancellation, or a fatal error, it will shutdown the threadpool, which will interrupt the background threads. In this case, we should avoid the noise of `Reporter.error` from the generic exception handlers around these tasks.

adriaanm commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -422,7 +423,7 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
       }
     }
-
+    SourceReader
```
stray?

retronym committed:
Also convert interruption during source file reading into cancellation

adriaanm commented:
I suggest we move this to 2.12.7. Let me know by Thursday if you disagree. Will need to drop the cancellation check in `typed1` in any case, and perhaps a few `if debug` guards.

retronym commented:
Delaying is okay with me. We need to fix it, but the problem is not visible
under default sbt settings which disables cancellation, so it isn’t worth
delaying 2.12.6

lrytz commented on src/compiler/scala/tools/nsc/io/SourceReader.scala:
```diff
@@ -38,7 +39,11 @@ class SourceReader(decoder: CharsetDecoder, reporter: Reporter) {
+    catch {
+      case ex: InterruptedException => throw ex
+      case _: ClosedByInterruptException => throw new InterruptedException
+      case e: Exception => reportEncodingError("" + file) ; Array()
```
Instead of adding this pattern in various places, could we just change `case _: Throwable` (or `case _: Exception`) to `case NonFatal(_)`? Should `ClosedByInterruptException` be added to the fatal list in `NonFatal`?

retronym replied:
Maybe, and maybe :)

But I'm not really game to change `NonFatal` itself (that could change user code, for the better or worse), and I'm also reluctant to make transition the compiler to using `NonFatal` in a piecemeal manner (we should do this refactoring holistically with some design and tests)

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/GeneratedClassHandler.scala:
```diff
@@ -153,6 +154,8 @@ private[jvm] object GeneratedClassHandler {
           unitInPostProcess.task.value.get.get
         } catch {
+          case _: ClosedByInterruptException => throw new InterruptedException()
+          case ex: InterruptedException => throw ex
```
This is redundant, `NonFatal` lets `InterruptedException` through

retronym committed:
Address review comments

adriaanm commented:
Seemingly unrelated failure -- rebase needed? https://scala-ci.typesafe.com/job/scala-2.12.x-validate-main/749/testReport/junit/test.files/run_/t2873_scala/

retronym commented:
rebased. 🤞

SethTisue commented:
sbt ticket on whether `true` should be the default for `cancelable`: sbt/sbt#3252

fommil commented:
@retronym if this also works in the PC, it could fix one of the biggest problems we faced in ENSIME and ScalaIDE, look for "One of the things that would really help is the ability to kill a hung PC instance": http://ensime.github.io/lsp-wg/#analyser
